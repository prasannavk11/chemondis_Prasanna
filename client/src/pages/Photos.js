import React, { useState, useEffect, useGlobal } from "reactn";
import { Link, useParams, useNavigate } from "react-router-dom";
import { getPhotos } from "../api";
import Grid from "../components/Grid";
import PageSize from "../components/PageSize";
import Pagination from "../components/Pagination";
import Loading from "../components/Loading";
import { getUserDetail, getAlbumDetail } from "../shared/utils";

const Photos = () => {
  const navigate = useNavigate();
  const { id } = useParams();
  const [users] = useGlobal("userList");
  const [albums] = useGlobal("albumList");
  const [photos, setPhotos] = useState([]);
  const [photosPageNumber, setPhotosPageNumber] = useState(0);
  const [photosPageSize, setPhotosPageSize] = useState(20);
  const [loading, setLoading] = useState(true);
  const [noMoreData, setNoMoreData] = useState(false);
  const [userData, setUserData] = useState("");
  const [albumData, setAlbumData] = useState({});
  const [error, setError] = useState(false);

  const albumDetail = () => {
    const tmpAlbumDetail = getAlbumDetail(albums, id);
    return tmpAlbumDetail;
  };

  const userDetail = () => {
    const tmpAlbumDetail = albumDetail();
    const tmpUserDetail = tmpAlbumDetail
      ? getUserDetail(users, tmpAlbumDetail.userId)
      : navigate("/");
    setUserData(tmpUserDetail);
    setAlbumData(tmpAlbumDetail);
  };

  const fetchPhotos = async () => {
    const responseJson = await getPhotos(
      id,
      photosPageNumber * photosPageSize,
      photosPageSize
    );
    if (responseJson) {
      setPhotos(responseJson);
      if (responseJson.length === 0) {
        setNoMoreData(true);
      }
    } else {
      setError(true);
    }
    setLoading(false);
  };

  useEffect(() => {
    setLoading(true);
    setNoMoreData(false);
    setError(false);
    fetchPhotos();
    userDetail();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [photosPageNumber, photosPageSize]);

  const onChangePageSize = (e) => {
    const tmpPageSize = parseInt(e.target.textContent);
    setPhotosPageNumber(0);
    setPhotosPageSize(tmpPageSize);
  };

  const onChangePageNumber = (e) => {
    const tmpPageNumber = e.target.textContent;
    tmpPageNumber === "Next"
      ? setPhotosPageNumber(photosPageNumber + 1)
      : setPhotosPageNumber(photosPageNumber - 1);
  };

  return (
    <section className="album-grid container">
      <Link to="/"> Go home</Link>
      <h1 className="my-4 text-center text-lg-left">{albumData && albumData.title}</h1>
      <h4 className="my-4 text-center text-lg-left">{userData}</h4>
      <PageSize
        sizes={[20, 30, 50]}
        count={photosPageSize}
        onChangePageSize={(e) => onChangePageSize(e)}
        disabled={loading || error}
      />
      <div className="row gallery">
        {loading && !noMoreData ? (
          <Loading />
        ) : (
          photos &&
          photos.map((photo, index) => {
            const tmpData = {
              ...photo,
              username: userData,
            };
            return <Grid key={index} type="photos" data={tmpData} />;
          })
        )}
        {noMoreData && "No more records found"}
        {error && "Error while fetching data. Kindly reload."}
      </div>
      <Pagination
        pageNumber={photosPageNumber}
        onChangePageNumber={(e) => onChangePageNumber(e)}
        disabled={loading || error}
        noMoreData={noMoreData || error}
      />
    </section>
  );
};

export default Photos;
