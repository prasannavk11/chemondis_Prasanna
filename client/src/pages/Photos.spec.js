import { MemoryRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import Photos from "./Photos";

afterEach(() => {
  jest.clearAllMocks();
  jest.resetAllMocks();
});

describe("Photos", () => {
  it("should render successfully", () => {
    const { baseElement } = render(<Photos />, {
      wrapper: MemoryRouter,
    });
    expect(baseElement).toMatchSnapshot();
  });
});
