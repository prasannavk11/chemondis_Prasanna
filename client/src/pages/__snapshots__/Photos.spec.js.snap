// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photos should render successfully 1`] = `
<body>
  <div>
    <section
      class="album-grid container"
    >
      <a
        href="/"
      >
         Go home
      </a>
      <h1
        class="my-4 text-center text-lg-left"
      />
      <h4
        class="my-4 text-center text-lg-left"
      />
      <nav
        aria-label="Page navigation example"
      >
        <ul
          class="pagination justify-content-end"
        >
          <li
            class="page-item disabled"
          >
            <div
              class="page-link cursor-pointer active"
            >
              20
            </div>
          </li>
          <li
            class="page-item disabled"
          >
            <div
              class="page-link cursor-pointer"
            >
              30
            </div>
          </li>
          <li
            class="page-item disabled"
          >
            <div
              class="page-link cursor-pointer"
            >
              50
            </div>
          </li>
        </ul>
      </nav>
      <div
        class="row gallery"
      >
        <div
          class="d-flex justify-content-center"
        >
          <div
            class="spinner-border"
            role="status"
          />
        </div>
      </div>
      <nav
        aria-label="Page navigation example"
      >
        <ul
          class="pagination justify-content-center"
        >
          <li
            class="page-item disabled"
          >
            <div
              class="page-link"
            >
              Previous
            </div>
          </li>
          <li
            class="page-item disabled"
          >
            <div
              class="page-link"
            >
              Next
            </div>
          </li>
        </ul>
      </nav>
    </section>
  </div>
</body>
`;
